### PR TITLE
P-256 ARMv8 optimization - first steps

### PR DESCRIPTION
**Pull request by nebeid**
### Issues:
Resolves CryptoAlg-536, CryptoAlg-556

### Description of changes: 
This PR is the first in a sequence of PR's that introduce ARMv8 assembly optimisations for P-256 into AWS-LC. The goal is bringing the P-256 performance on ARMv8 at par with x86_64. The ARMv8 assembly code is taken from OpenSSL 1.1.1 (at [this commit](https://github.com/openssl/openssl/commit/46a9ee8c796c8b5f8d95290676119b4f3d72be91)). 
The goal is achieved by reusing the AWS-LC P-256 implementation in `p256-x86_64.c` (nistz256) with ARMv8. This is possible because the assembly functions required to support that code have their analogous functions in the imported OpenSSL. Namely, the file [openss/crypto/ec/asm/ecp_nistz256-armv8.pl](https://github.com/openssl/openssl/blob/46a9ee8c796c8b5f8d95290676119b4f3d72be91/crypto/ec/asm/ecp_nistz256-armv8.pl) was imported with slight modification in the first 2 commits.
However, there are 3 x86_64 assembly functions that do not have corresponding functions in that file.
Those functions are `ecp_nistz256_select_w5` and `ecp_nistz256_select_w7` and `beeu_mod_inverse_vartime`. 
The 3rd commit in this PR implements: `ecp_nistz256_select_w5`.

### Call-outs:
- The code duplication of C functions from `p256-x86_64.c` into `p256.c` will be removed in subsequent commits; when all assembly functions called by nistz256 are available (or worked around, if needed). Instead the file `p256-x86_64.c` will be renamed (possibly `p256-nistz256.c`) and will be used with both x86_64 and armv8 platforms.

### Testing:
Tested `boringssl/crypto/crypto_test` on ARMv8 and x86_64 platforms.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: patches/0013-start-arm-optimized-nistz256 [hunk 1]
```diff
@@ -0,0 +1,2577 @@
...
++                                            ////    movdqu	$Rc, 16*2($val)
++                                            ////    movdqu	$Rd, 16*3($val)
++                                            ////    movdqu	$Re, 16*4($val)
++                                            ////    movdqu	$Rf, 16*5($val)
++	ret
++.size	ecp_nistz256_select_w5,.-ecp_nistz256_select_w5
++___
++}
++########################################################################
++# scatter-gather subroutines
```
Comment (torben-hansen) [inline patches/0013-start-arm-optimized-nistz256:2334]: Can't we remove these functions? They are being replaced by the "select" constant-time functions instead and nothin in AWS-LC should be calling these functions.

File: patches/0013-start-arm-optimized-nistz256 [hunk 2]
```diff
@@ -0,0 +1,2577 @@
...
+===================================================================
+--- aws-lc.orig/third_party/boringssl/crypto/fipsmodule/ec/ec.c
++++ aws-lc/third_party/boringssl/crypto/fipsmodule/ec/ec.c
+@@ -249,6 +249,9 @@ DEFINE_METHOD_FUNCTION(struct built_in_c
+ #if !defined(OPENSSL_NO_ASM) && defined(OPENSSL_X86_64) && \
+     !defined(OPENSSL_SMALL)
+       EC_GFp_nistz256_method();
++#elif !defined(OPENSSL_NO_ASM) && defined(OPENSSL_AARCH64) && \
++    !defined(OPENSSL_SMALL)
++      EC_GFp_nistz256_arm_method();
```
Comment (torben-hansen) [inline patches/0013-start-arm-optimized-nistz256:31]: Should we name the x86_64 EC_METHOD for `EC_GFp_nistz256_x86_64_method`?

**Review by torben-hansen — COMMENTED** [review]
To test my understanding, is this a correct description of the purpose of this PR?
This PR is the first in a sequence of PR's that is including ARMv8 assembly optimisations for P256 into AWS-LC. The ARMv8 assembly code is taken from OpenSSL 1.1.1. The goal is to reuse the AWS-LC P256 implementation in `p256-x86_64.c`. This is achieved by making the assembly functions for ARMv8 match the x86-64 assembly functions. The initial missing ARMv8 functions are `beeu_mod_inverse_vartime`, `ecp_nistz256_select_w5` and `ecp_nistz256_select_w7`. This PR also implements: `ecp_nistz256_select_w5`.

I have manually verified the following:
* The replicated C-functions are the same as in the file `p256-x86_64.c`
* The ARMv8 code taken from OpenSSL 1.1.1 (`p256-armv8-asm.pl`) is identical to the code imported here, except for 1) modifying some paths 2) including `"openssl/arm_arch.h"` 3) implementation of `ecp_nistz256_select_w5`.

I have the following general questions and remarks:
* There are test cases in `p256-x86-64-test.cc` that tests some of the assembly functions and the p256 in general. These should also be run for ARM (currently disabled for anything else than x86_64). What is the testing plan for this? What is the testing plan for the new assembly functions and the changes in general? We should enable this as early as possible.
* What are the differences between scatter-gather functions and select functions? Is one obviously worse than the other in terms of resistance against various side-channel attack vectors?
* For good measure: did you verify that the assembly functions are compiled and actually run on the ARM instance?
* For good measure: did you run the clang linter?

**Comment by nebeid** [comment]
Thanks @torben-hansen's for the great comment, I edited the PR message.
In reply to your questions
> I have the following general questions and remarks:
> 
>     * There are test cases in `p256-x86-64-test.cc` that tests some of the assembly functions 
>        and the p256 in general. These should also be run for ARM (currently disabled for anything 
>        else than x86_64). What is the testing plan for this? 
>        What is the testing plan for the new assembly functions and the changes in general? 
>        We should enable this as early as possible.
Great call! They are now enabled in [the last commit, cf9e88f](https://github.com/awslabs/aws-lc/pull/52/commits/cf9e88f9c8836283b0afdfd0393c04fe3034cb28). Only the tests of the 2 functions that are to be implemented are disabled.

>     * What are the differences between scatter-gather functions and select functions? 
>        Is one obviously worse than the other in terms of resistance against various side-channel attack vectors?
I didn't go in details but there are attacks on the scatter/gather techniques for RSA, [CacheBleed, Yarom etal., CHES-2016](https://link.springer.com/chapter/10.1007/978-3-662-53140-2_17). It relies on measuring time variations due to contention on the same cache bank. The attack was carried on OpenSSL v. 1.0.2f. It's plausible that OpenSSL improved the resistance of memory access to this type of attacks.
On the other hand, the select method was what was originally in the [patch provided by Gueron and Krasnov to OpenSSL](https://rt.openssl.org/Ticket/Display.html?id=3149&user=guest&pass=guest) and was adopted by BoringSSL since. I believe it performs better on platforms with SIMD instructions. In fact, in the patch comments, it is mentioned that OpenSSL preferred to keep the scatter/gather technique because the select technique was less performant on platforms without SIMD.

>     * For good measure: did you verify that the assembly functions are compiled and actually run on the ARM instance?
> 
Yes, on an ARMv8 M6g instance and an x86_64 MacBook.

>     * For good measure: did you run the clang linter?
I did on files under `ec`. It outputs a lot of warnings for existent code. We can discuss further the course of action.

File: patches/series [hunk 3]
```diff
@@ -10,3 +10,4 @@
 0010-reduce-threads-num-in-rsa-test.patch
 0011-fix-fuzz-test-path-to-libc.patch
 0012-replace-boringssl-api-version-macro.patch
+0013-start-arm-optimized-nistz256
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/series]: Done

File: patches/0013-start-arm-optimized-nistz256 [hunk 4]
```diff
@@ -0,0 +1,2577 @@
...
+Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/internal.h
+===================================================================
+--- aws-lc.orig/third_party/boringssl/crypto/fipsmodule/ec/internal.h
++++ aws-lc/third_party/boringssl/crypto/fipsmodule/ec/internal.h
+@@ -713,6 +713,10 @@ const EC_METHOD *EC_GFp_nistp256_method(
+ // x86-64 optimized P256. See http://eprint.iacr.org/2013/816.
+ const EC_METHOD *EC_GFp_nistz256_method(void);
+ 
++// [TODO] EC_GFp_nistz256_arm_method is a GFp method using montgomery multiplication, with
++// x86-64 optimized P256. See http://eprint.iacr.org/2013/816.
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-start-arm-optimized-nistz256]: Done. Edited.

File: patches/0013-start-arm-optimized-nistz256 [hunk 5]
```diff
@@ -0,0 +1,2577 @@
...
+ // P-256 field operations.
+Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256.c
+===================================================================
+--- aws-lc.orig/third_party/boringssl/crypto/fipsmodule/ec/p256.c
++++ aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256.c
+@@ -24,6 +24,12 @@
+ #include <openssl/mem.h>
+ #include <openssl/type_check.h>
+ 
++#if !defined(OPENSSL_NO_ASM) && defined(OPENSSL_AARCH64) &&	\
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-start-arm-optimized-nistz256]: Done. Thanks. It's one ifdef now.

File: patches/0013-start-arm-optimized-nistz256 [hunk 6]
```diff
@@ -0,0 +1,2577 @@
...
++  OPENSSL_memcpy(a.X, a_->X.words, P256_LIMBS * sizeof(BN_ULONG));
++  OPENSSL_memcpy(a.Y, a_->Y.words, P256_LIMBS * sizeof(BN_ULONG));
++  OPENSSL_memcpy(a.Z, a_->Z.words, P256_LIMBS * sizeof(BN_ULONG));
++  ecp_nistz256_point_double(&a, &a);
++  OPENSSL_memcpy(r->X.words, a.X, P256_LIMBS * sizeof(BN_ULONG));
++  OPENSSL_memcpy(r->Y.words, a.Y, P256_LIMBS * sizeof(BN_ULONG));
++  OPENSSL_memcpy(r->Z.words, a.Z, P256_LIMBS * sizeof(BN_ULONG));
++}
++
++typedef P256_POINT_AFFINE PRECOMP256_ROW[64];
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-start-arm-optimized-nistz256]: Done. Thanks.

File: patches/0013-start-arm-optimized-nistz256 [hunk 7]
```diff
@@ -0,0 +1,2577 @@
...
++
++my ($acc6,$acc7)=($ap,$bp);	# used in __ecp_nistz256_sqr_mont
++
++$code.=<<___;
++#include "openssl/arm_arch.h"
++
++.text
++___
++########################################################################
++# Convert ecp_nistz256_table.c to layout expected by ecp_nistz_gather_w7
```
[the remark this answers is not in the document]
Comment (nebeid) [reply patches/0013-start-arm-optimized-nistz256]: They are removed. Thanks.

File: patches/0013-start-arm-optimized-nistz256 (hunk 1 above)
Comment (nebeid) [reply patches/0013-start-arm-optimized-nistz256]: They are removed. Thanks.

File: patches/0013-start-arm-optimized-nistz256 (hunk 2 above)
Comment (nebeid) [reply patches/0013-start-arm-optimized-nistz256]: We should end up with this method being used both with x86_64 and ARMv8

**Review by torben-hansen — APPROVED** [review]
I have good confidence in the functional correctness of the new assembly function. But I'm not familiar enough with ARMv8 to verify that the function is constant-time, e.g. does the instructions used operate in constant-time and how does our supported OS's/platforms factor into that? Given the cryptographic sensitivity of this code I propose we do the following before merging into Main:
- More thoroughly review the new ARMv8 code
- Apply fuzz testing since the code is handling sensitive user-supplied input

**Review by andrew-kaufman — APPROVED** [review]

**Pull request merged**